Change ALL tests that rely on internet services to use mock services

When offline, all these tests fail:

- [x] FAILED tests/match/test_GT_matching.py::test_genotype_matching - assert 0 == 2
- [x] FAILED tests/match/test_matching_handler.py::test_internal_matching - assert 0 == 2
- [x] FAILED tests/match/test_matching_handler.py::test_internal_matching_with_threshold - assert 0 == 2
- [x] FAILED tests/server/test_server_responses.py::test_add_patient - AssertionError: assert False
- [x] FAILED tests/server/test_server_responses.py::test_match_hgnc_symbol_patient - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Ma...
- [x] FAILED tests/server/test_server_responses.py::test_match_ensembl_patient - AssertionError: assert False
- [x] FAILED tests/server/test_server_responses.py::test_match_entrez_patient - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max ret...
- [x] FAILED tests/utils/test_ensembl_rest_api.py::test_ping_ensemble_37 - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='rest.ensembl.org', port=443): Max retries exceede...
- [x] FAILED tests/utils/test_ensembl_rest_api.py::test_ping_ensemble_38 - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='rest.ensembl.org', port=443): Max retries exceede...
- [x] FAILED tests/utils/test_ensembl_rest_api.py::test_send_gene_request - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max retries...
- [x] FAILED tests/utils/test_ensembl_rest_api.py::test_send_request_wrong_url - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max re...
- [x] FAILED tests/utils/test_gene.py::test_entrez_to_symbol - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max retries exceeded wit...
- [x] FAILED tests/utils/test_variant.py::test_liftover_37_38 - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max retries exceeded wi...
- [x] FAILED tests/utils/test_variant.py::test_liftover_38_37 - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max retries exceeded wi...
- [x] FAILED tests/utils/test_variant.py::test_liftover_MT_variant - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max retries exceed...
- [x] FAILED tests/utils/test_variant.py::test_liftover_bad_request - requests.exceptions.ConnectionError: HTTPSConnectionPool(host='grch37.rest.ensembl.org', port=443): Max retries excee...

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,7 +1,7 @@
 ## [] -
 ### Changed
 - Speed-up CI tests by caching installation of libs and splitting tests into randomized groups using pytest-test-groups
-- Mock Ensembl services (gene conversion and liftover) used in patient parsing tests
+- Mock Ensembl services (gene conversions and liftover) used all tests
 - Mock the download of files containing phenotype resources in tests
 
 ## [3.0] - 2021-12-27

--- a/tests/match/test_GT_matching.py
+++ b/tests/match/test_GT_matching.py
@@ -1,11 +1,43 @@
 # -*- coding: utf-8 -*-
-
-from patientMatcher.parse.patient import mme_patient
+import responses
 from patientMatcher.match.genotype_matcher import match
+from patientMatcher.parse.patient import mme_patient
 
 
-def test_genotype_matching(database, gpx4_patients):
+@responses.activate
+def test_genotype_matching(database, gpx4_patients, mock_symbol_2_ensembl):
     """Testing the genotyping matching algorithm"""
+
+    # GIVEN a mocked Ensembl REST API:
+    for hgnc_symbol, ensembl_id in mock_symbol_2_ensembl.items():
+        responses.add(
+            responses.GET,
+            f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/{hgnc_symbol}?external_db=HGNC",
+            json=[{"id": ensembl_id}],
+            status=200,
+        )
+
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/lookup/id/ENSG00000167468",
+        json=[{"display_name": "GPX4"}],
+        status=200,
+    )
+
+    # GIVEN a mocked liftover service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1105813..1105814/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1106232..1106238/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
 
     # load 2 test patients in mock database
     for patient in gpx4_patients:

--- a/tests/match/test_GT_matching.py
+++ b/tests/match/test_GT_matching.py
@@ -5,17 +5,16 @@ from patientMatcher.parse.patient import mme_patient
 
 
 @responses.activate
-def test_genotype_matching(database, gpx4_patients, mock_symbol_2_ensembl):
+def test_genotype_matching(database, gpx4_patients):
     """Testing the genotyping matching algorithm"""
 
     # GIVEN a mocked Ensembl REST API:
-    for hgnc_symbol, ensembl_id in mock_symbol_2_ensembl.items():
-        responses.add(
-            responses.GET,
-            f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/{hgnc_symbol}?external_db=HGNC",
-            json=[{"id": ensembl_id}],
-            status=200,
-        )
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
+        json=[{"id": "ENSG00000167468", "type": "gene"}],
+        status=200,
+    )
 
     # GIVEN a mocked Ensembl gene lookup service:
     responses.add(

--- a/tests/match/test_matching_handler.py
+++ b/tests/match/test_matching_handler.py
@@ -62,7 +62,39 @@ def test_internal_matching(mock_app, database, gpx4_patients):
     assert higest_scored_patient["score"]["patient"] > lowest_scored_patient["score"]["patient"]
 
 
-def test_internal_matching_with_threshold(database, gpx4_patients):
+@responses.activate
+def test_internal_matching_with_threshold(mock_app, database, gpx4_patients):
+
+    # GIVEN a mocked Ensembl REST API:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
+        json=[{"id": "ENSG00000167468", "type": "gene"}],
+        status=200,
+    )
+
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/lookup/id/ENSG00000167468",
+        json=[{"display_name": "GPX4"}],
+        status=200,
+    )
+
+    # GIVEN a mocked liftover service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1105813..1105814/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1106232..1106238/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
+
     # load 2 test patients in mock database
     for patient in gpx4_patients:
         mme_pat = mme_patient(patient, True)  # convert gene symbol to ensembl

--- a/tests/match/test_matching_handler.py
+++ b/tests/match/test_matching_handler.py
@@ -1,12 +1,44 @@
 # -*- coding: utf-8 -*-
 import requests
+import responses
 from patientMatcher.match.handler import external_matcher, internal_matcher, save_async_response
 from patientMatcher.parse.patient import mme_patient
 from patientMatcher.utils.add import backend_add_patient
 
 
-def test_internal_matching(database, gpx4_patients):
+@responses.activate
+def test_internal_matching(mock_app, database, gpx4_patients):
     """Testing the combined matching algorithm"""
+
+    # GIVEN a mocked Ensembl REST API:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
+        json=[{"id": "ENSG00000167468", "type": "gene"}],
+        status=200,
+    )
+
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/lookup/id/ENSG00000167468",
+        json=[{"display_name": "GPX4"}],
+        status=200,
+    )
+
+    # GIVEN a mocked liftover service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1105813..1105814/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1106232..1106238/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
 
     # load 2 test patients in mock database
     for patient in gpx4_patients:

--- a/tests/server/test_server_responses.py
+++ b/tests/server/test_server_responses.py
@@ -624,8 +624,39 @@ def test_match_ensembl_patient(mock_app, test_client, gpx4_patients, database):
     assert match["data"]["patient"]["genomicFeatures"][0]["gene"]["_geneName"] == "GPX4"
 
 
+@responses.activate
 def test_match_entrez_patient(mock_app, test_client, gpx4_patients, database):
     """Test matching patient with ensembl gene against patientMatcher database (internal matching)"""
+
+    # GIVEN a mocked Ensembl REST API converting gene symbol to Ensembl ID
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
+        json=[{"id": "ENSG00000167468", "type": "gene"}],
+        status=200,
+    )
+
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/lookup/id/ENSG00000167468",
+        json=[{"display_name": "GPX4"}],
+        status=200,
+    )
+
+    # GIVEN a mocked liftover service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1105813..1105814/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1106232..1106238/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
 
     # add an authorized client to database
     ok_token = test_client["auth_token"]

--- a/tests/server/test_server_responses.py
+++ b/tests/server/test_server_responses.py
@@ -547,8 +547,39 @@ def test_match_hgnc_symbol_patient(mock_app, gpx4_patients, test_client, databas
             assert pat["score"]["patient"] > 0
 
 
+@responses.activate
 def test_match_ensembl_patient(mock_app, test_client, gpx4_patients, database):
     """Test matching patient with ensembl gene against patientMatcher database (internal matching)"""
+
+    # GIVEN a mocked Ensembl REST API converting gene symbol to Ensembl ID
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
+        json=[{"id": "ENSG00000167468", "type": "gene"}],
+        status=200,
+    )
+
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/lookup/id/ENSG00000167468",
+        json=[{"display_name": "GPX4"}],
+        status=200,
+    )
+
+    # GIVEN a mocked liftover service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1105813..1105814/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1106232..1106238/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
 
     # add an authorized client to database
     ok_token = test_client["auth_token"]

--- a/tests/server/test_server_responses.py
+++ b/tests/server/test_server_responses.py
@@ -195,7 +195,7 @@ def test_add_patient_malformed_data(mock_app, test_client, gpx4_patients, test_n
 def test_add_patient(mock_app, test_client, gpx4_patients, test_node, database):
     """Test adding a patient by sending a POST request to the add endpoint with valid data"""
 
-    # GIVEN a mocked Ensembl REST API:
+    # GIVEN a mocked Ensembl REST API converting gene symbol to Ensembl ID
     responses.add(
         responses.GET,
         f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
@@ -468,8 +468,39 @@ def test_patient_matches(mock_app, database, match_objs, test_client):
     assert len(matches) == 2
 
 
+@responses.activate
 def test_match_hgnc_symbol_patient(mock_app, gpx4_patients, test_client, database):
-    """Testing matching patient with gene symbl against patientMatcher database (internal matching)"""
+    """Testing matching patient with gene symbol against patientMatcher database (internal matching)"""
+
+    # GIVEN a mocked Ensembl REST API converting gene symbol to Ensembl ID
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
+        json=[{"id": "ENSG00000167468", "type": "gene"}],
+        status=200,
+    )
+
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/lookup/id/ENSG00000167468",
+        json=[{"display_name": "GPX4"}],
+        status=200,
+    )
+
+    # GIVEN a mocked liftover service:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1105813..1105814/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/19:1106232..1106238/GRCh38?content-type=application/json",
+        json=[],
+        status=200,
+    )
 
     # add an authorized client to database
     ok_token = test_client["auth_token"]

--- a/tests/server/test_server_responses.py
+++ b/tests/server/test_server_responses.py
@@ -1,6 +1,7 @@
 # -*- coding: utf-8 -*-
 import json
 
+import responses
 from patientMatcher.__version__ import __version__
 from patientMatcher.auth.auth import authorize
 from patientMatcher.match.handler import patient_matches
@@ -190,8 +191,17 @@ def test_add_patient_malformed_data(mock_app, test_client, gpx4_patients, test_n
     assert response.status_code == 422
 
 
+@responses.activate
 def test_add_patient(mock_app, test_client, gpx4_patients, test_node, database):
     """Test adding a patient by sending a POST request to the add endpoint with valid data"""
+
+    # GIVEN a mocked Ensembl REST API:
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/GPX4?external_db=HGNC",
+        json=[{"id": "ENSG00000167468", "type": "gene"}],
+        status=200,
+    )
 
     patient_data = gpx4_patients[1]
 

--- a/tests/server/test_server_responses.py
+++ b/tests/server/test_server_responses.py
@@ -3,7 +3,6 @@ import json
 
 import responses
 from patientMatcher.__version__ import __version__
-from patientMatcher.auth.auth import authorize
 from patientMatcher.match.handler import patient_matches
 from patientMatcher.parse.patient import mme_patient
 from patientMatcher.server.controllers import validate_response

--- a/tests/utils/test_ensembl_rest_api.py
+++ b/tests/utils/test_ensembl_rest_api.py
@@ -1,58 +1,86 @@
 # -*- coding: UTF-8 -*-
 import pytest
+import responses
 from patientMatcher.utils import ensembl_rest_client as ensembl_api
 from requests.exceptions import MissingSchema
 
 
+@responses.activate
 def test_except_on_invalid_response():
     """Test function that creates exception with message when response returned from Ensembl REST API has status code !=200"""
 
     # GIVEN a response from Ensembl service with status != 200
+    responses.add(
+        responses.GET,
+        f"https://rest.ensembl.org//info/ping",
+        json={"ping": 1},
+        status=500,
+    )
     client = ensembl_api.EnsemblRestApiClient()
-
-    class MockResponse:
-        def __init__(self):
-            self.status = 500
-
-    mockresp = MockResponse()
 
     # Then it should trigger an exception
     with pytest.raises(Exception):
         result = client.except_on_invalid_response(mockresp)
 
 
-def test_ping_ensemble_37():
-    """Test ping ensembl server containing human build 37"""
-    client = ensembl_api.EnsemblRestApiClient()
-    data = client.ping_server()
-    assert data == {"ping": 1}
+@responses.activate
+def test_ping_ensembl():
+    """Test ping ensembl server"""
 
+    # GIVEN a mocked Ensembl REST API server
+    responses.add(
+        responses.GET,
+        f"https://rest.ensembl.org//info/ping",
+        json={"ping": 1},
+        status=200,
+    )
 
-def test_ping_ensemble_38():
-    """Test ping ensembl server containing human build 38"""
     client = ensembl_api.EnsemblRestApiClient(build="38")
     data = client.ping_server()
     assert data == {"ping": 1}
 
 
+@responses.activate
 def test_send_gene_request():
     """Test send request with correct params and endpoint"""
+
+    # GIVEN a gene with Ensembl ID
     url = "https://grch37.rest.ensembl.org/lookup/id/ENSG00000103591"
+
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        url,
+        json={"display_name": "AAGAB"},
+        status=200,
+    )
+
     client = ensembl_api.EnsemblRestApiClient()
     data = client.send_request(url)
+
     # get info for the ensembl gene
     assert data["display_name"] == "AAGAB"
 
 
+@responses.activate
 def test_send_request_wrong_url():
     """Successful requests are tested by other tests in this file.
     This test will trigger errors instead.
     """
+
     not_an_url = "foo"
+
     client = ensembl_api.EnsemblRestApiClient()
     data = client.send_request(not_an_url)
     assert type(data) == MissingSchema
 
     url = f"https://grch37.rest.ensembl.org/{not_an_url}"
+    # GIVEN a mocked Ensembl gene lookup service:
+    responses.add(
+        responses.GET,
+        url,
+        json={"error": "page not found"},
+        status=404,
+    )
     data = client.send_request(url)
     assert type(data) == ValueError

--- a/tests/utils/test_gene.py
+++ b/tests/utils/test_gene.py
@@ -1,32 +1,31 @@
 # -*- coding: utf-8 -*-
+import responses
 from patientMatcher.utils.ensembl_rest_client import requests
 from patientMatcher.utils.gene import ensembl_to_symbol, entrez_to_symbol, symbol_to_ensembl
 
 
+@responses.activate
 def test_entrez_to_symbol():
     """Test the function converting entrez ID to gene symbol"""
-
     # GIVEN an entrez ID
     entrez_id = "3735"
     # THAT should return a symbol
     symbol = "KARS"
 
-    # GIVEN a patched API response
-    class MockResponse(object):
-        def __init__(self):
-            self.status_code = 200
-
-        def json(self):
-            return {"display_id": symbol}
-
-    def mock_get(url, headers):
-        return MockResponse()
+    # GIVEN a mocked Ensembl REST API converting entrez ID to gene symbol
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/name/human/{entrez_id}?external_db=EntrezGene",
+        json=[{"display_id": symbol}],
+        status=200,
+    )
 
     # The EnsemblRestApiClient should return the right Ensembl ID
     assert entrez_to_symbol(entrez_id) == symbol
 
 
-def test_symbol_to_ensembl_one_ensembl_id(monkeypatch):
+@responses.activate
+def test_symbol_to_ensembl_one_ensembl_id():
     """
     Test function converting official gene symbol to ensembl ID using the Ensembl REST API
     Test case when the Ensembl API return only one Ensembl ID as result
@@ -34,20 +33,16 @@ def test_symbol_to_ensembl_one_ensembl_id(monkeypatch):
 
     # GIVEN a gene symbol
     hgnc_symbol = "AAGAB"
+    # That should return an Ensembl ID
     ensembl_id = "ENSG00000103591"
 
-    # GIVEN a patched API response
-    class MockResponse(object):
-        def __init__(self):
-            self.status_code = 200
-
-        def json(self):
-            return [{"id": ensembl_id, "type": "gene"}]
-
-    def mock_get(url, headers):
-        return MockResponse()
-
-    monkeypatch.setattr(requests, "get", mock_get)
+    # GIVEN a mocked Ensembl REST API converting gene symbol to Ensembl ID
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/xrefs/symbol/homo_sapiens/{hgnc_symbol}?external_db=HGNC",
+        json=[{"id": "ENSG00000103591", "type": "gene"}],
+        status=200,
+    )
 
     # The EnsemblRestApiClient should return the right Ensembl ID
     assert ensembl_id == symbol_to_ensembl(hgnc_symbol)
@@ -59,6 +54,7 @@ def test_symbol_to_ensembl_multiple_ensembl_id(monkeypatch):
     """
     # GIVEN a gene symbol
     hgnc_symbol = "SKI"
+    # Corresponfing to 2 different ensembl IDs
     rigth_ensembl_id = "ENSG00000157933"
     wrong_ensembl_id = "ENSG00000054392"
 
@@ -107,6 +103,7 @@ def test_symbol_to_ensembl_multiple_ensembl_id(monkeypatch):
     assert ensembl_id == rigth_ensembl_id
 
 
+@responses.activate
 def test_ensembl_to_symbol(monkeypatch):
     """Test converting ensembl ID to official gene symbol using the Ensembl APIs"""
 

--- a/tests/utils/test_variant.py
+++ b/tests/utils/test_variant.py
@@ -1,6 +1,8 @@
+import responses
 from patientMatcher.utils.variant import liftover
 
 
+@responses.activate
 def test_liftover_37_38():
     """Test variant liftover from GRCh37 to GRCh38"""
 
@@ -8,6 +10,35 @@ def test_liftover_37_38():
     chromosome = "X"
     start = 1000000
     end = 1000000
+
+    # GIVEN a patched liftover Ensembl service
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/{chromosome}:{start}..{end}/GRCh38?content-type=application/json",
+        json={
+            "mappings": [
+                {
+                    "original": {
+                        "assembly": "GRCh37",
+                        "end": 1000000,
+                        "coord_system": "chromosome",
+                        "seq_region_name": "X",
+                        "start": 1000000,
+                        "strand": 1,
+                    },
+                    "mapped": {
+                        "strand": 1,
+                        "start": 1039265,
+                        "seq_region_name": "X",
+                        "end": 1039265,
+                        "coord_system": "chromosome",
+                        "assembly": "GRCh38",
+                    },
+                }
+            ]
+        },
+        status=200,
+    )
 
     # THEN the service should return valid mappings
     mappings = liftover("GRCh37", chromosome, start, end)
@@ -22,6 +53,7 @@ def test_liftover_37_38():
     assert mappings[0]["mapped"]["assembly"] == "GRCh38"
 
 
+@responses.activate
 def test_liftover_38_37():
     """Test variant liftover from GRCh38 to GRCh37"""
 
@@ -29,6 +61,53 @@ def test_liftover_38_37():
     chromosome = "X"
     start = 1039265
     end = 1039265
+
+    # GIVEN a patched liftover Ensembl service
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh38/{chromosome}:{start}..{end}/GRCh37?content-type=application/json",
+        json={
+            "mappings": [
+                {
+                    "original": {
+                        "seq_region_name": "X",
+                        "end": 1039265,
+                        "assembly": "GRCh38",
+                        "coord_system": "chromosome",
+                        "strand": 1,
+                        "start": 1039265,
+                    },
+                    "mapped": {
+                        "seq_region_name": "X",
+                        "end": 1000000,
+                        "assembly": "GRCh37",
+                        "strand": 1,
+                        "coord_system": "chromosome",
+                        "start": 1000000,
+                    },
+                },
+                {
+                    "original": {
+                        "assembly": "GRCh38",
+                        "seq_region_name": "X",
+                        "end": 1039265,
+                        "start": 1039265,
+                        "strand": 1,
+                        "coord_system": "chromosome",
+                    },
+                    "mapped": {
+                        "seq_region_name": "HG480_HG481_PATCH",
+                        "end": 1000000,
+                        "assembly": "GRCh37",
+                        "strand": 1,
+                        "coord_system": "chromosome",
+                        "start": 1000000,
+                    },
+                },
+            ]
+        },
+        status=200,
+    )
 
     # THEN the service should return valid mappings
     mappings = liftover("GRCh38", chromosome, start, end)
@@ -44,6 +123,7 @@ def test_liftover_38_37():
     assert mappings[0]["mapped"]["assembly"] == "GRCh37"
 
 
+@responses.activate
 def test_liftover_MT_variant():
     """Test liftover for a mitochondrial variant from GRCh37 to GRCh38"""
 
@@ -51,6 +131,35 @@ def test_liftover_MT_variant():
     chromosome = "MT"
     start = 7278
     end = 7278
+
+    # GIVEN a patched liftover Ensembl service
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/{chromosome}:{start}..{end}/GRCh38?content-type=application/json",
+        json={
+            "mappings": [
+                {
+                    "mapped": {
+                        "strand": 1,
+                        "end": 7278,
+                        "assembly": "GRCh38",
+                        "coord_system": "chromosome",
+                        "seq_region_name": "MT",
+                        "start": 7278,
+                    },
+                    "original": {
+                        "strand": 1,
+                        "seq_region_name": "MT",
+                        "start": 7278,
+                        "end": 7278,
+                        "assembly": "GRCh37",
+                        "coord_system": "chromosome",
+                    },
+                }
+            ]
+        },
+        status=200,
+    )
 
     # THEN the service should return valid mappings
     mappings = liftover("GRCh37", chromosome, start, end)
@@ -65,6 +174,7 @@ def test_liftover_MT_variant():
     assert mappings[0]["mapped"]["assembly"] == "GRCh38"
 
 
+@responses.activate
 def test_liftover_bad_request():
     """Test variant liftover with non-valid request params"""
 
@@ -72,6 +182,14 @@ def test_liftover_bad_request():
     chromosome = "M"
     start = 7278
     end = 7278
+
+    # GIVEN a patched liftover Ensembl service
+    responses.add(
+        responses.GET,
+        f"https://grch37.rest.ensembl.org/map/human/GRCh37/{chromosome}:{start}..{end}/GRCh38?content-type=application/json",
+        json={"error": "'No slice found for M on coord_system chromosome and assembly GRCh37'"},
+        status=200,
+    )
 
     # THEN the service should not return mappings
     mappings = liftover("GRCh37", chromosome, start, end)

